# it is a regration problem

## we are doing the machine learning on house price data set convert itno markdown

<table align="center" width=100%>
<tr>
    </td>
         </td>
     </td> 
     <td>
        <div align="center">
        <font color="#21618C" size=8px>
         <b> Regression modeling on House price Data 
        </b>
     </font>
    </div>
 </td>
</tr>
</table>
         

# Price Data 

# Problem Statement

#This dataset is created for the prediction of House Prices from an Indian perspective Towns. The dataset contains various features that are important during the application for the selling the Houses. The predicted output obtained from the classification algorithm gives a fair idea about the chances of a Houses Sold.

# import 'pandas'  panda is utilize to format the data structure it can convert any data structure into pandas dataframe.
import pandas as pd
# import 'numpy'  to find stratestical summary we use numpy
import numpy as np

#import subpackage of Matplotlib
import matplotlib.pyplot as plt 
from matplotlib.colors import ListedColormap

#import 'Seaborn
import seaborn as sns

#to suppress warnings from warnings
import filtervarnings. filtersarnings('ignore')




### Load the dataset

In [9]:
# Load the & name it df_houseprice
import pandas as pd

In [10]:
df_houseprice = pd.read_csv('C:\\Users\\sai\\Downloads\\house_price.csv')

In [11]:
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

# Data overview

In [13]:
# Name the columns
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

# Interpretetion

 - Price :Price of the House.
 - resid_area :Residential Area where my house is located.

In [15]:
# basic info of the Dataset
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        506 non-null    float64
 1   resid_area   506 non-null    float64
 2   air_qual     506 non-null    float64
 3   room_num     506 non-null    float64
 4   age          506 non-null    float64
 5   distance     506 non-null    float64
 6   teachers     506 non-null    float64
 7   poor_prop    506 non-null    float64
 8   airport      506 non-null    object 
 9   n_hos_beds   498 non-null    float64
 10  n_hot_rooms  506 non-null    float64
 11  waterbody    351 non-null    object 
 12  rainfall     506 non-null    int64  
 13  bus_ter      506 non-null    object 
 14  parks        506 non-null    float64
 15  Sold         506 non-null    int64  
dtypes: float64(11), int64(2), object(3)
memory usage: 63.4+ KB


# Interpretation

- we have 13 numerical columns & 3 categorical column
- we have NULL values in the data
- we have total 16 columns in the data

In [18]:
# shape of the data 
df_houseprice.shape

(506, 16)

# interpretation

- we have total 16 columns
- we have 506 observations

In [21]:
# summarize the basic statistics of Data 
df_houseprice.describe().T

count       mean        std        min        25%        50%  \
price        506.0  22.528854   9.182176   5.000000  17.025000  21.200000   
resid_area   506.0  41.136779   6.860353  30.460000  35.190000  39.690000   
air_qual     506.0   0.554695   0.115878   0.385000   0.449000   0.538000   
room_num     506.0   6.284634   0.702617   3.561000   5.885500   6.208500   
age          506.0  68.574901  28.148861   2.900000  45.025000  77.500000   
distance     506.0   3.971996   2.108532   1.130000   2.270000   3.385000   
teachers     506.0  21.544466   2.164946  18.000000  19.800000  20.950000   
poor_prop    506.0  12.653063   7.141062   1.730000   6.950000  11.360000   
n_hos_beds   498.0   7.899767   1.476683   5.268000   6.634500   7.999000   
n_hot_rooms  506.0  13.041605   5.238957  10.057600  11.189800  12.720000   
rainfall     506.0  39.181818  12.513697   3.000000  28.000000  39.000000   
parks        506.0   0.054454   0.010632   0.033292   0.046464   0.053507   
Sold         506.0   0.454545   0.498422   0.000000   0.000000   0.000000   

                   75%         max  
price        25.000000   50.000000  
resid_area   48.100000   57.740000  
air_qual      0.624000    0.871000  
room_num      6.623500    8.780000  
age          94.075000  100.000000  
distance      5.367500   12.320000  
teachers     22.600000   27.400000  
poor_prop    16.955000   37.970000  
n_hos_beds    9.088000   10.876000  
n_hot_rooms  14.170800  101.120000  
rainfall     50.000000   60.000000  
parks         0.061397    0.086711  
Sold          1.000000    1.000000

# interpretation

- we have outliers in data,most of the outliers are in AGE column

### data preprocessing

### a.data clinging

In [26]:
#show me first five observation
df_houseprice.head(5)

                   
                   

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

In [27]:
# rename  to the columns
#step 1 : columns (' show the columns names')
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

In [28]:
#step 2 : allocate new columns names to old columns names
# make the dicitionary('old columns nmae': 'new columns name')

d1={'price': 'Price',
'resid_aera': 'Residental_Area',
'air_quality':'Air_Quality',
'room_num': 'Room_Number',
'age': 'Age',
'distance': 'Distance',
'teachers': 'Teachers', 
'poor_prop': 'proportion_of_poor_population',
'airport': 'Airport', 
'n_hos_beds': 'Number_of_Hospital_Beds',
'n_hot_rooms': 'Tumber_of_Hotel_Rooms',
'waterbody':'Waterbody',
'rainfall': 'Rainfall',
'bus_ter': 'Bus_Terminal',
'Parks': 'Parks',
}




## b. Null Value Handling

In [30]:
df_houseprice=df_houseprice.rename(columns=d1)
df_houseprice.head(5)

Price  resid_area  air_qual  Room_Number   Age  Distance  Teachers  \
0   24.0       32.31     0.538        6.575  65.2      4.35      24.7   
1   21.6       37.07     0.469        6.421  78.9      4.99      22.2   
2   34.7       37.07     0.469        7.185  61.1      5.03      22.2   
3   33.4       32.18     0.458        6.998  45.8      6.21      21.3   
4   36.2       32.18     0.458        7.147  54.2      6.16      21.3   

   proportion_of_poor_population Airport  Number_of_Hospital_Beds  \
0                           4.98     YES                    5.480   
1                           9.14      NO                    7.332   
2                           4.03      NO                    7.394   
3                           2.94     YES                    9.268   
4                           5.33      NO                    8.824   

   Tumber_of_Hotel_Rooms Waterbody  Rainfall Bus_Terminal     parks  Sold  
0                11.1920     River        23          YES  0.049347     0  
1                12.1728      Lake        42          YES  0.046146     1  
2               101.1200       NaN        38          YES  0.045764     0  
3                11.2672      Lake        45          YES  0.047151     0  
4                11.2896      Lake        55          YES  0.039474     0

In [31]:
# Authenticate the data
df_houseprice.columns

Index(['Price', 'resid_area', 'air_qual', 'Room_Number', 'Age', 'Distance',
       'Teachers', 'proportion_of_poor_population', 'Airport',
       'Number_of_Hospital_Beds', 'Tumber_of_Hotel_Rooms', 'Waterbody',
       'Rainfall', 'Bus_Terminal', 'parks', 'Sold'],
      dtype='object')

Laws that we need to followwhile processing the null values(specially for manufacturing)
- when we have less than 15% data is missing-> remove the observations
-betwwen 20-70% data is missing->impute the null values
-greater than 75% values are missing->remove the column

In [33]:
# find the value(null value)
df_houseprice.isnull().sum()

Price                              0
resid_area                         0
air_qual                           0
Room_Number                        0
Age                                0
Distance                           0
Teachers                           0
proportion_of_poor_population      0
Airport                            0
Number_of_Hospital_Beds            8
Tumber_of_Hotel_Rooms              0
Waterbody                        155
Rainfall                           0
Bus_Terminal                       0
parks                              0
Sold                               0
dtype: int64

In [34]:
# percentage missing values
round(((df_houseprice.isnull().sum())/len(df_houseprice))*100,3)

Price                             0.000
resid_area                        0.000
air_qual                          0.000
Room_Number                       0.000
Age                               0.000
Distance                          0.000
Teachers                          0.000
proportion_of_poor_population     0.000
Airport                           0.000
Number_of_Hospital_Beds           1.581
Tumber_of_Hotel_Rooms             0.000
Waterbody                        30.632
Rainfall                          0.000
Bus_Terminal                      0.000
parks                             0.000
Sold                              0.000
dtype: float64

In [35]:
# find the nature of the column
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Price                          506 non-null    float64
 1   resid_area                     506 non-null    float64
 2   air_qual                       506 non-null    float64
 3   Room_Number                    506 non-null    float64
 4   Age                            506 non-null    float64
 5   Distance                       506 non-null    float64
 6   Teachers                       506 non-null    float64
 7   proportion_of_poor_population  506 non-null    float64
 8   Airport                        506 non-null    object 
 9   Number_of_Hospital_Beds        498 non-null    float64
 10  Tumber_of_Hotel_Rooms          506 non-null    float64
 11  Waterbody                      351 non-null    object 
 12  Rainfall                       506 non-null    int

In [36]:
#this is a categorical column so we can replace the null with the mode
df_houseprice.Waterbody.mode()

0    River
Name: Waterbody, dtype: object

In [37]:
#replace the null
df_houseprice.Waterbody=df_houseprice.Waterbody.fillna('River')

In [38]:
#percentage misssing values
round(((df_houseprice.isnull().sum())/len(df_houseprice))*100,3)

Price                            0.000
resid_area                       0.000
air_qual                         0.000
Room_Number                      0.000
Age                              0.000
Distance                         0.000
Teachers                         0.000
proportion_of_poor_population    0.000
Airport                          0.000
Number_of_Hospital_Beds          1.581
Tumber_of_Hotel_Rooms            0.000
Waterbody                        0.000
Rainfall                         0.000
Bus_Terminal                     0.000
parks                            0.000
Sold                             0.000
dtype: float64

In [39]:
#in Number_of_Hospital_Beds contain less than 10% missing values
df_houseprice.dropna(inplace=True)

In [40]:
round(((df_houseprice.isnull().sum())/len(df_houseprice))*100.3)

Price                            0.0
resid_area                       0.0
air_qual                         0.0
Room_Number                      0.0
Age                              0.0
Distance                         0.0
Teachers                         0.0
proportion_of_poor_population    0.0
Airport                          0.0
Number_of_Hospital_Beds          0.0
Tumber_of_Hotel_Rooms            0.0
Waterbody                        0.0
Rainfall                         0.0
Bus_Terminal                     0.0
parks                            0.0
Sold                             0.0
dtype: float64

## prepare data for machine learning 

In [42]:
target=df_houseprice['Sold']

In [43]:
df_independent=df_houseprice.drop('Sold' , axis=1)

## Separate the categorical columns from independent variable

In [45]:
df_cat=df_independent.select_dtypes(include='object')

In [46]:
df_cat.head()

Airport Waterbody Bus_Terminal
0     YES     River          YES
1      NO      Lake          YES
2      NO     River          YES
3     YES      Lake          YES
4      NO      Lake          YES

In [47]:
#encode the categorical variable
encoded= pd.get_dummies(df_cat)

In [48]:
encoded.head()

Airport_NO  Airport_YES  Waterbody_Lake  Waterbody_Lake and River  \
0       False         True           False                     False   
1        True        False            True                     False   
2        True        False           False                     False   
3       False         True            True                     False   
4        True        False            True                     False   

   Waterbody_River  Bus_Terminal_YES  
0             True              True  
1            False              True  
2             True              True  
3            False              True  
4            False              True

## Scale the numerical data

In [50]:
df_num = df_independent.select_dtypes(include='number')

In [51]:
from sklearn.preprocessing import StandardScaler

In [52]:
ss=StandardScaler()

In [53]:
scale=ss.fit_transform(df_num)
scale

array([[ 0.15479189, -1.28244435, -0.13709359, ..., -0.35211111,
        -1.30206786, -0.47313205],
       [-0.10659165, -0.59005882, -0.73319492, ..., -0.16601943,
         0.21519397, -0.77522463],
       [ 1.32012686, -0.59005882, -0.73319492, ..., 16.71034024,
        -0.10422957, -0.81123666],
       ...,
       [ 0.14390091,  0.11687263,  0.16527665, ..., -0.16252832,
        -0.66322077,  0.30293265],
       [-0.06302773,  0.11687263,  0.16527665, ...,  0.40379146,
         0.6144734 ,  0.59750894],
       [-0.38975716,  0.11687263,  0.16527665, ..., -0.54943507,
         0.45476163,  0.56369225]])

In [54]:
df_scale=pd.DataFrame(scale,columns=df_num.columns)
df_scale

Price  resid_area  air_qual  Room_Number       Age  Distance  \
0    0.154792   -1.282444 -0.137094     0.408975 -0.115644  0.172845   
1   -0.106592   -0.590059 -0.733195     0.189974  0.369741  0.476100   
2    1.320127   -0.590059 -0.733195     1.276446 -0.260905  0.495053   
3    1.178544   -1.301354 -0.828226     1.010516 -0.802977  1.054179   
4    1.483492   -1.301354 -0.828226     1.222407 -0.505369  1.030487   
..        ...         ...       ...          ...       ...       ...   
493 -0.019464    0.116873  0.165277     0.434572  0.022531 -0.637413   
494 -0.215501    0.116873  0.165277    -0.238073  0.291796 -0.732180   
495  0.143901    0.116873  0.165277     0.979230  0.798438 -0.779564   
496 -0.063028    0.116873  0.165277     0.720411  0.738208 -0.684797   
497 -0.389757    0.116873  0.165277    -0.366061  0.437057 -0.599506   

     Teachers  proportion_of_poor_population  Number_of_Hospital_Beds  \
0    1.462167                      -1.066368                -1.640299   
1    0.304963                      -0.484136                -0.384875   
2    0.304963                      -1.199330                -0.342847   
3   -0.111631                      -1.351886                 0.927490   
4   -0.111631                      -1.017382                 0.626514   
..        ...                            ...                      ...   
493 -1.176258                      -0.409958                 0.981720   
494 -1.176258                      -0.492534                -0.872945   
495 -1.176258                      -0.973995                -1.641654   
496 -1.176258                      -0.856429                 0.027273   
497 -1.176258                      -0.660486                 1.613499   

     Tumber_of_Hotel_Rooms  Rainfall     parks  
0                -0.352111 -1.302068 -0.473132  
1                -0.166019  0.215194 -0.775225  
2                16.710340 -0.104230 -0.811237  
3                -0.337843  0.454762 -0.680402  
4                -0.333593  1.253320 -1.404723  
..                     ...       ...       ...  
493              -0.164805 -0.982644  0.155116  
494               0.022197 -1.541636  0.522886  
495              -0.162528 -0.663221  0.302933  
496               0.403791  0.614473  0.597509  
497              -0.549435  0.454762  0.563692  

[498 rows x 12 columns]

In [55]:
df_scale=df_scale.reset_index()
df_scale

index     Price  resid_area  air_qual  Room_Number       Age  Distance  \
0        0  0.154792   -1.282444 -0.137094     0.408975 -0.115644  0.172845   
1        1 -0.106592   -0.590059 -0.733195     0.189974  0.369741  0.476100   
2        2  1.320127   -0.590059 -0.733195     1.276446 -0.260905  0.495053   
3        3  1.178544   -1.301354 -0.828226     1.010516 -0.802977  1.054179   
4        4  1.483492   -1.301354 -0.828226     1.222407 -0.505369  1.030487   
..     ...       ...         ...       ...          ...       ...       ...   
493    493 -0.019464    0.116873  0.165277     0.434572  0.022531 -0.637413   
494    494 -0.215501    0.116873  0.165277    -0.238073  0.291796 -0.732180   
495    495  0.143901    0.116873  0.165277     0.979230  0.798438 -0.779564   
496    496 -0.063028    0.116873  0.165277     0.720411  0.738208 -0.684797   
497    497 -0.389757    0.116873  0.165277    -0.366061  0.437057 -0.599506   

     Teachers  proportion_of_poor_population  Number_of_Hospital_Beds  \
0    1.462167                      -1.066368                -1.640299   
1    0.304963                      -0.484136                -0.384875   
2    0.304963                      -1.199330                -0.342847   
3   -0.111631                      -1.351886                 0.927490   
4   -0.111631                      -1.017382                 0.626514   
..        ...                            ...                      ...   
493 -1.176258                      -0.409958                 0.981720   
494 -1.176258                      -0.492534                -0.872945   
495 -1.176258                      -0.973995                -1.641654   
496 -1.176258                      -0.856429                 0.027273   
497 -1.176258                      -0.660486                 1.613499   

     Tumber_of_Hotel_Rooms  Rainfall     parks  
0                -0.352111 -1.302068 -0.473132  
1                -0.166019  0.215194 -0.775225  
2                16.710340 -0.104230 -0.811237  
3                -0.337843  0.454762 -0.680402  
4                -0.333593  1.253320 -1.404723  
..                     ...       ...       ...  
493              -0.164805 -0.982644  0.155116  
494               0.022197 -1.541636  0.522886  
495              -0.162528 -0.663221  0.302933  
496               0.403791  0.614473  0.597509  
497              -0.549435  0.454762  0.563692  

[498 rows x 13 columns]

In [56]:
x=pd.concat([df_num,encoded],axis=1)
x.head()

Price  resid_area  air_qual  Room_Number   Age  Distance  Teachers  \
0   24.0       32.31     0.538        6.575  65.2      4.35      24.7   
1   21.6       37.07     0.469        6.421  78.9      4.99      22.2   
2   34.7       37.07     0.469        7.185  61.1      5.03      22.2   
3   33.4       32.18     0.458        6.998  45.8      6.21      21.3   
4   36.2       32.18     0.458        7.147  54.2      6.16      21.3   

   proportion_of_poor_population  Number_of_Hospital_Beds  \
0                           4.98                    5.480   
1                           9.14                    7.332   
2                           4.03                    7.394   
3                           2.94                    9.268   
4                           5.33                    8.824   

   Tumber_of_Hotel_Rooms  Rainfall     parks  Airport_NO  Airport_YES  \
0                11.1920        23  0.049347       False         True   
1                12.1728        42  0.046146        True        False   
2               101.1200        38  0.045764        True        False   
3                11.2672        45  0.047151       False         True   
4                11.2896        55  0.039474        True        False   

   Waterbody_Lake  Waterbody_Lake and River  Waterbody_River  Bus_Terminal_YES  
0           False                     False             True              True  
1            True                     False            False              True  
2           False                     False             True              True  
3            True                     False            False              True  
4            True                     False            False              True

In [64]:
# import train-test data split

from sklearn.model_selection import train_test_split
# add a constant columto the dataframe 
# while using the 'logit' method in the statsmodels library the method do not
# we can add the intercept to the set of independent variables 'add.constant'
import statsmodels
import statsmodels.api as sm
x=sm.add_constant(x)
x_train, x_test, y_train, y_test = train_test_split(x,target,random_state=10,test_size=0.2)

In [58]:
import statsmodels

# decision tree classifier

In [67]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
 #assuming x_train 
tree_model = DecisionTreeClassifier()

tree_model.fit(x_train , y_train)

prediction = tree_model.predict(x_test)



In [71]:
prediction

array([0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0,
       1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0,
       1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0], dtype=int64)

In [73]:
accuracy=accuracy_score(y_test,prediction)
print("accuracy:",accuracy)

accuracy: 0.61
